aptos-move/move-examples/defi/aptoswap/core/sources/liquidity_pool.move

```diff
@@ -0,0 +1,442 @@
+
+        // mint new lp tokens(fa)
+        fungible_asset::mint(&pool.lp_token_mint_ref, provided_liq)
+    }
+
+    /// Burn liquidity coins (LP) and get back X and Y coins from reserves.
+    /// * `lp tokens` - LP tokens to burn.
+    /// Returns both X and Y coins - `(Coin<X>, Coin<Y>)`.
+    public fun burn(
+        lp_tokens: FungibleAsset
+    ): (FungibleAsset, FungibleAsset) acquires LiquidityPool {
+        let lp_token_metadata = fungible_asset::metadata_from_asset(&lp_tokens);
+        let lp = object::convert<Metadata, LiquidityPool>(lp_token_metadata);
+        let pool = borrow_global<LiquidityPool>(object::object_address(&lp));
+
+        let amount_to_burn = fungible_asset::amount(&lp_tokens);
+        let lp_token_supply = lp_token_supply(pool);
+        let x_store = *vector::borrow(&pool.store_pair, 0);
+        let y_store = *vector::borrow(&pool.store_pair, 1);
+        let x_reserve_size = fungible_asset::balance(x_store);
+        let y_reserve_size = fungible_asset::balance(y_store);
+
+        // Compute x, y coin values for provided lp_coins value
+        let x_amount_to_redeem = (math128::mul_div(
+            (amount_to_burn as u128),
+            (x_reserve_size as u128),
+            lp_token_supply
+        ) as u64);
+        let y_amount_to_redeem = (math128::mul_div(
+            (amount_to_burn as u128),
+            (y_reserve_size as u128),
+            lp_token_supply
+        ) as u64);
+        assert!(x_amount_to_redeem != 0 && y_amount_to_redeem != 0, ERR_NOT_ENOUGH_LP_TOKEN_TO_BURN);
+
+        let pool_signer = generate_signer_for_extending(&pool.extend_ref);
+        // Withdraw those values from reserves
+        let x_coin = fungible_asset::withdraw(&pool_signer, x_store, x_amount_to_redeem);
+        let y_coin = fungible_asset::withdraw(&pool_signer, y_store, y_amount_to_redeem);
+
+        fungible_asset::burn(&pool.lp_token_burn_ref, lp_tokens);
+
+        (x_coin, y_coin)
+    }
+
+    /// Swap a for b.
+    /// * `in` - fungible asset a to swap.
+    /// * `to` - the type of b.
+    /// * `to_amount` - expected amount of the coins of b to get out.
+    /// return `to_amount` of b as fungible asset.
+    public fun swap(
+        in: FungibleAsset,
+        to: Object<Metadata>,
+        to_amount: u64,
+    ): FungibleAsset acquires LiquidityPool {
+        let from = fungible_asset::metadata_from_asset(&in);
+        let (from_index, pool) = if (is_valid_pair(from, to)) {
+            (0, borrow_lp(from, to))
+        } else {
+            (1, borrow_lp(to, from))
+        };
+        let from_reserve = fungible_asset::balance(*vector::borrow(&pool.store_pair, from_index));
+        let to_reserve = fungible_asset::balance(*vector::borrow(&pool.store_pair, 1 - from_index));
+        let k_before_swap = (from_reserve as u256) * (to_reserve as u256) * (FEE_SCALE * FEE_SCALE as u256);
+        let (out, k_after_swap_and_fee) = swap_and_calculate_new_k(pool, from_index, in, to_amount);
```

Comment:
This doesn't seem right. What is to_amount for? Technically, to can be calculated by maintaining the product k, so to_amount is not needed. The only reason you'd often these in DEXes is for users to provide a min_out as a way to indicate max slippage tolerated (if slippage is too high, you can get a lot less back than expected so providing a min guarantees the tx would fail if you receive less than that)